app/lib/three_scale/spam_protection/protector.rb

```diff
@@ -89,6 +89,33 @@ def to_str
       end
 
       alias to_s to_str
+
+      private
+
+      def request_session
+        return {} if template.controller.blank?
+
+        template.controller.request.session
+      end
+
+      def spam_post_request?
+        !http_method.get? && considered_as_spam?
+      end
+
+      def spam_get_request?
+        http_method.get? && was_considered_as_spam?
+      end
```

Comment:
At least for me, I think these the whole thing could be simpler to read like the following:

```ruby
def captcha_needed?
  captcha_required? || possible_spam?
end

private

def possible_spam?
  if http_method.get?
    marked_as_possible_spam?
  else
    mark_possible_spam!
  end
end

def marked_as_possible_spam?
  request_session[:marked_as_possible_spam_until].to_i > Time.now.utc.to_i
end

def mark_possible_spam!
  return unless protector.spam?
  request_session[:marked_as_possible_spam_until] = (Time.now.utc + 5.minutes).to_i
end
```